### PR TITLE
HPCC-25230 Set ldapUser to access file for cloud environment

### PR DESCRIPTION
**Pull request by wangkx**
1. Add the ldapUser setting for roxie file access.
2. Set the ldapUser setting in containerized environment.
3. Remove the roxieUsers from the QueryFilesInUse.
4. Move QueryFilesInUse::updateUsers() from ws_workunitsService.hpp
to ws_workunitsServiceQuerySets.cpp.
5. In QueryFilesInUse::updateUsers(), add the code to get the ldapUser
setting and set the roxieUserMap.
7. In CRoxieDaliHelper, read the ldapUser setting and set the userdesc.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck I just added the roxie support for the configured secret. Please review. 
@jakesmith do we still need both user name and password when authorizing the file access in dali (request from roxie or esp)?

**Comment by jakesmith** [comment]
>@jakesmith do we still need both user name and password when authorizing the file access in dali (request from roxie or esp)?

No. The internal clients are implicitly trusted to have the correct user, the user name determines access rights though..
But I'm not sure if the workunit token is still used somewhere.
@ghalliday and/or @RussWhitehead - may be able to comment more on this.

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 1]
```diff
@@ -420,6 +420,64 @@ void QueryFilesInUse::loadTargets(IPropertyTree *t, unsigned flags)
...
+                throw makeStringExceptionV(-1, "Error extracting username/password for FileSecretKey %s.", secretKey.str());
+        }
+
+        Owned<IUserDescriptor> user = createUserDescriptor();
+        if (!username.isEmpty())
+            user->set(username, password);
+        else
+            user->set("roxie", "");
+        roxieUserMap.setValue(target.str(), user);
+        roxieUsers.append(*user.getClear());
```
Comment (jakesmith) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp:462]: minor: maybe not new, but "roxieUsers" does not appear to be used by anything - if so, should be removed.

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 2]
```diff
@@ -420,6 +420,64 @@ void QueryFilesInUse::loadTargets(IPropertyTree *t, unsigned flags)
...
+            Owned<IPropertyTree> secretTree;
+            if (vaultId.length() > 0)
+                secretTree.setown(getVaultSecret("authn", vaultId.str(), secretKey.str(), nullptr));
+            else
+                secretTree.setown(getSecret("authn", secretKey.str()));
+            if (!secretTree)
+                throw makeStringExceptionV(-1, "Error retrieving Roxie file secret for FileSecretKey %s.", secretKey.str());
+
+            getSecretKeyValue(username, secretTree, "username");
+            getSecretKeyValue(password, secretTree, "password");
```
Comment (jakesmith) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp:451]: the password is not going to be used, the roxie (or any dali client, and in near future esp client with certificate) is implicitly trusted.

**Review by jakesmith — COMMENTED** [review]
@wangkx - I'm not convinced we want a secret for this, but others may know better.

Roxie needs to be configured with a user to use, but not a user/pw combination.
Internal clients are implicitly trusted to be using the correct user.

**Comment by afishbeck** [comment]
@jakesmith I agree that the password shouldn't be required, and that if that's the case we don't need a secret.  Our offline discussion about that is why Kevin asked.  
If it's just the username (likely a system username I would think, like 'roxie1_file_access') then I think having it directly in the values configuration is fine.

**Comment by wangkx** [comment]
I will change to a username setting in values.yaml.

**Review by afishbeck — COMMENTED** [review]
@wangkx looks good.

**Review by afishbeck — APPROVED** [review]

**Comment by wangkx** [comment]
@richardkchapman please review and merge.

**Comment by richardkchapman** [comment]
Needs rebasing for a clean merge

**Comment by wangkx** [comment]
@richardkchapman rebased.

**Review by ghalliday — COMMENTED** [review]
The change looks reasonable.  One question about the whole change is why was a new systemUser chosen instead of using ldapUser (which I think provides exactly the same information in bare-metal).
It would avoid the new functions in the interface, and it would allow some of the other code to be commoned up.

**Comment by wangkx** [comment]
> systemUser

**Comment by wangkx** [comment]
@ghalliday I am using systemUser because the name ldapUser may confuse people. The systemUser may be used not only for LDAP. If you want, I can change it to use the existing methods, like getLdapUser().

**Comment by wangkx** [comment]
@ghalliday The systemUser is renamed to ldapUser. Please review.

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 3]
```diff
@@ -420,6 +420,36 @@ void QueryFilesInUse::loadTargets(IPropertyTree *t, unsigned flags)
     }
 }
 
+void QueryFilesInUse::updateUsers()
+{
+    SCMStringBuffer target;
+
+#ifdef _CONTAINERIZED
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp:427]: This code is now essentially the same for container/non container, and should be simplified.

File: helm/hpcc/templates/_helpers.tpl [hunk 4]
```diff
@@ -808,6 +808,7 @@ Generate instance queue names
   prefix: {{ .prefix }}
   {{- end }}
   queriesOnly: true
+  ldapUser: {{ .ldapUser }}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:811]: @richardkchapman what would be your preferred name for this option.  Kevin had systemUser.  It isn't really a system user or ldap user (security may not use ldap).  I think it is the user that is used for all file lookups from roxie.  I don't have a strong opinion any way.

**Review by ghalliday — COMMENTED** [review]
@wangkx that looks cleaner.  One comment about simplifying the code, and a question for Richard.

File: esp/smc/SMCLib/TpContainer.cpp [hunk 5]
```diff
@@ -598,7 +600,7 @@ class CContainerWUClusterInfo : public CSimpleInterfaceOf<IConstWUClusterInfo>
     }
     virtual const char *getLdapUser() const override
     {
-        UNIMPLEMENTED;
+        return ldapUser.get();
     }
     virtual const char *getLdapPassword() const override
     {
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/smc/SMCLib/TpContainer.cpp:606]: Will do

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp (hunk 3 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsQuerySets.cpp:427]: Will do

File: helm/hpcc/templates/_helpers.tpl (hunk 4 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:811]: It's called ldapUser on bare metal legacy configs, I think. So probably makes sense to use the same name?

**Comment by wangkx** [comment]
@ghalliday please review my changes.

**Review by ghalliday — APPROVED** [review]
Looks good

**Comment by ghalliday** [comment]
@wangkx please squash

**Comment by wangkx** [comment]
@ghalliday @richardkchapman this PR is ready to be merged.

**Pull request merged**